coordinator/src/main/scala/filodb.coordinator/queryplanner/SingleClusterPlanner.scala

```diff
@@ -269,45 +304,143 @@ class SingleClusterPlanner(val dataset: Dataset,
-      case lp: ScalarBinaryOperation       => materializeScalarBinaryOperation(qContext, lp)
-      case lp: SubqueryWithWindowing       => materializeSubqueryWithWindowing(qContext, lp)
-      case lp: TopLevelSubquery            => materializeTopLevelSubquery(qContext, lp)
-      case lp: LabelCardinality            => materializeLabelCardinality(qContext, lp)
+      case lp: ApplyAbsentFunction         => materializeAbsentFunction(qContext, lp, forceInProcess)
+      case lp: ApplyLimitFunction          => materializeLimitFunction(qContext, lp, forceInProcess)
+      case lp: ScalarBinaryOperation       => materializeScalarBinaryOperation(qContext, lp, forceInProcess)
+      case lp: SubqueryWithWindowing       => materializeSubqueryWithWindowing(qContext, lp, forceInProcess)
+      case lp: TopLevelSubquery            => materializeTopLevelSubquery(qContext, lp, forceInProcess)
+      case lp: LabelCardinality            => materializeLabelCardinality(qContext, lp, forceInProcess)
       //case _                               => throw new BadQueryException("Invalid logical plan")
     }
   }
   // scalastyle:on cyclomatic.complexity
 
-  override def materializeBinaryJoin(qContext: QueryContext,
-                                     lp: BinaryJoin): PlanResult = {
-    val lhs = walkLogicalPlanTree(lp.lhs, qContext)
+  /**
+   * Returns the shards spanned by a LogicalPlan's data.
+   * @return an occupied Option iff shards could be determined for all leaf-level plans.
+   */
+  private def getShardSpanFromLp(lp: LogicalPlan,
+                                 qContext: QueryContext): Option[Set[Int]] = {
+
+    case class LeafInfo(lp: LogicalPlan,
+                        filters: Set[ColumnFilter],
+                        tslabels: Option[Seq[String]])
+
+    // construct a LeafInfo for each leaf
+    val leaves = findLeafLogicalPlans(lp).map{ leafPlan =>
+      val filters = getColumnFilterGroup(leafPlan)
+      assert(filters.size == 1, s"expected leaf plan to yield single filter group, but got ${filters.size}")
+      val targetSchemaProvider = qContext.plannerParams.targetSchemaProvider
+      if (targetSchemaProvider.isDefined) {
+        val tsFunc = targetSchemaProvider.get.targetSchemaFunc(filters.head.toSeq)
+        val (start, stop) = leafPlan match {
+          // TODO(a_theimer): other leaf plan types? Range selector types?
+          case rs: RawSeries => rs.rangeSelector match {
+            case is: IntervalSelector => (is.from, is.to)
+            case _ => throw new IllegalArgumentException(s"unhandled type: ${rs.rangeSelector.getClass}")
+          }
+          case _ => throw new IllegalArgumentException(s"unhandled type: ${leafPlan.getClass}")
+        }
+        val tsLabels = findTargetSchema(tsFunc, start, stop).map(_.schema)
+        LeafInfo(leafPlan, filters.head, tsLabels)
+      } else {
+        LeafInfo(leafPlan, filters.head, None)
+      }
+    }
```

Comment:
I'll try to make this logic less confusing (and, at the very least, not check that the same TSP is defined for each plan).